opensearch data stream compatibility in fluent-bit

adiforluls opened the pull request:
<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" --> 
Addresses #7079 

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [ ] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

adiforluls commented:
Sample configuration file
```
[Output]
    Name    opensearch
    Match    kube.*
    Host    foo.bar
    Port    9200
    Suppress_Type_Name    true
    Data_Stream_Mode    true
    Data_Stream_Name    test
    Template_File    /fluent-bit/etc/template.json
```
Debug logs from output
```
[2023/05/12 06:16:41] [debug] [output:opensearch:opensearch.0] create index template=0
[2023/05/12 06:16:41] [debug] [output:opensearch:opensearch.0] HTTP Status=200 URI=/_bulk
[2023/05/12 06:16:41] [debug] [output:opensearch:opensearch.0] OpenSearch response
```

adiforluls commented:
Inspired by and loosely based on https://github.com/fluent/fluent-plugin-opensearch#configuration---opensearch-output-data-stream

I'll be looking to add support for creating a default index template such that supplying an index template isn't necessary as a follow up if that's okay.

adiforluls commented:
Let me know the process for documentation changes as well (I see the `docs required` label, if I need to submit those changes hand in hand or after this PR goes in)

adiforluls commented:
Further looping in @cosmo0920 who I see wrote the fluentd plugin

leonardo-albertovich commented:
@cosmo0920 will make an primary review of the plugin focusing on the logical and practical viewpoint as time permits and I will do a secondary review focusing on coding style and lower level bugs once that's done.

This process will take place within the following week.

Thank you very much for taking the time to make this contribution.

PettitWesley commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -1089,6 +1224,27 @@ static struct flb_config_map config_map[] = {
+	 "Enable data stream mode"
+	},
+	{
+		FLB_CONFIG_MAP_STR, "data_stream_name", FLB_OS_DEFAULT_INDEX,
+		0, FLB_TRUE, offsetof(struct flb_opensearch, data_stream_name),
+		"Set the name of the target data stream"
+	},
+	{
+		FLB_CONFIG_MAP_STR, "data_stream_template_name", FLB_OS_DEFAULT_TEMPLATE,
+		0, FLB_TRUE, offsetof(struct flb_opensearch, data_stream_template_name),
+			"Set name of the index template corresponding to the data stream"
```
extra indenting here not present on other options

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -853,12 +900,88 @@ static int opensearch_error_check(struct flb_opensearch *ctx,
                                 struct flb_config *config)
 {
     int ret = -1;
+	int res;
```
Indentation with white spaces is better to use instead of tabs.

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -875,6 +998,18 @@ static void cb_opensearch_flush(struct flb_event_chunk *event_chunk,
+	*/
+
+	if (ctx->data_stream_mode == FLB_TRUE) {
+		res = create_index_template(ctx);
```
Putting on every request of flush is insufficient to handle on this.
We have to search the existing template if possible and not to overwrite by default.

adiforluls replied:
I'll look to update the logic, will introduce a `Template_Overwrite` flag in fluent-bit opensearch config which will be set to false by default. Now, if a template with the given name exists, overwrite only if the flag is set to true in the output config. Wdyt?

cosmo0920 replied:
Yup. Your understanding is correct. Thanks for your understanding.

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -473,13 +520,13 @@ static int opensearch_format(struct flb_config *config,
             }
         }
-        else if (ctx->current_time_index == FLB_TRUE) {
+		else if (ctx->current_time_index == FLB_TRUE) {
```
Did you use tabs instead of white spaces for indentations?
If so, we recommend to use white spaces for that purpose.

adiforluls replied:
Yeah I was using tabs. I'll fix this.

cosmo0920 reviewed (changes requested):
It's nice work for adding data streams compatibility

Just putting template file on every request is insufficient to handle data stream.
If possible, we need to check the existence of template files.
And, I noticed that tabs should be used in your patch of indentations.
Using white spaces should be better for that purpose.

leonardo-albertovich commented:
Just to add to what @cosmo0920 said, converting those tabs to white spaces is not just better, it's necessary to take this forward.

adiforluls commented:
cc: @cosmo0920 @leonardo-albertovich

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -853,12 +900,149 @@ static int opensearch_error_check(struct flb_opensearch *ctx,
+	 * retry if the file path is not set or if the template is missing
+	 */
+	if (ctx->template_file) {
+        if (ctx->template_overwrite == FLB_TRUE) {
+            return put_index_template(ctx);
+        }
+        else {
+            res = template_exists(ctx);
+            if (res == FLB_OS_SUCCESS) {
+                flb_plg_debug(ctx->ins, "template with given name %s exists", ctx->data_stream_template_name);
+                return FLB_OS_SUCCESS;
+            }
+            else if (res == FLB_OS_NOT_FOUND) {
+                flb_plg_debug(ctx->ins, "template with given name %s does not exist, creating it now", ctx->data_stream_template_name);
+                return put_index_template(ctx);
+            }
+            else {
+                flb_plg_error(ctx->ins, "error checking if the template exists, retrying");
+                return FLB_OS_RETRY;
+            }
+        }
```
Indentation with white spaces is needed for these lines.

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -1089,6 +1285,32 @@ static struct flb_config_map config_map[] = {
      "AWS Service Name"
     },
 #endif
+	/* Data stream compatibility */
+	{
+	 FLB_CONFIG_MAP_BOOL, "data_stream_mode", "false",
+	 0, FLB_TRUE, offsetof(struct flb_opensearch, data_stream_mode),
+	 "Enable data stream mode"
+	},
+	{
+     FLB_CONFIG_MAP_STR, "data_stream_name", FLB_OS_DEFAULT_INDEX,
+     0, FLB_TRUE, offsetof(struct flb_opensearch, data_stream_name),
+     "Set the name of the target data stream"
+	},
+    {
+     FLB_CONFIG_MAP_STR, "data_stream_template_name", FLB_OS_DEFAULT_TEMPLATE,
+     0, FLB_TRUE, offsetof(struct flb_opensearch, data_stream_template_name),
+     "Set name of the index template corresponding to the data stream"
+	},
+	{
+	 FLB_CONFIG_MAP_STR, "template_file", NULL,
+	 0, FLB_TRUE, offsetof(struct flb_opensearch, template_file),
+	 "Location of the index template file"
+	},
```
Indentation glitches are existing here. We need to fix the glitch.

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -875,6 +1059,18 @@ static void cb_opensearch_flush(struct flb_event_chunk *event_chunk,
         FLB_OUTPUT_RETURN(FLB_RETRY);
     }
 
+	/*
+	 * create index template if data stream mode is enabled
+	*/
+
+	if (ctx->data_stream_mode == FLB_TRUE) {
+		res = create_index_template(ctx);
+		flb_plg_debug(ctx->ins, "create index template=%i", res);
+		if (res != 0) {
+			FLB_OUTPUT_RETURN(FLB_RETRY);
+		}
+	}
+
```
8 spaces for indentation is not correct for our coding style. Could you use 4 spaces indentations?

adiforluls commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -1231,20 +1292,25 @@ static struct flb_config_map config_map[] = {
-		"Set the name of the target data stream"
+     FLB_CONFIG_MAP_STR, "data_stream_name", FLB_OS_DEFAULT_INDEX,
+     0, FLB_TRUE, offsetof(struct flb_opensearch, data_stream_name),
+     "Set the name of the target data stream"
```
Still need to fix the indentation I see, fixing in a follow up commit

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -336,39 +338,48 @@ static int opensearch_format(struct flb_config *config,
-            flb_sds_destroy(j_index);
-            return -1;
-        }
+	if (((ctx->data_stream_mode == FLB_TRUE && ctx->ra_data_stream == NULL) || (ctx->logstash_format == FLB_FALSE && ctx->ra_index == NULL)) && ctx->generate_id == FLB_FALSE) {
```
Break a long line, please.

cosmo0920 reviewed (commented):
I confirmed that the checking for the existence of template file is nice.
However, still indentation issue is persist.
Could you use 4-spaces indentation instead of 8-spaces when you use white spaces?

adiforluls commented:
Sure, it's my IDE's doing, will fix shortly

adiforluls commented:
> I confirmed that the checking for the existence of template file is nice. However, still indentation issue is persist. Could you use 4-spaces indentation instead of 8-spaces when you use white spaces?

Sure, it's my IDE's doing, will fix shortly

adiforluls commented:
@cosmo0920 @leonardo-albertovich I think I've addressed all the comments and fixed indentation issues. Let me know if more changes are needed. 

Once this goes in, I'll try and squeeze in the logic to create a default index template if the template file isn't provided as a follow up to this.

cosmo0920 commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -146,6 +157,8 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
         snprintf(ctx->uri, sizeof(ctx->uri) - 1, "%s/_bulk", path);
     }
 
+	snprintf(ctx->template_uri, sizeof(ctx->template_uri) - 1, "%s/_index_template/%s", path, ctx->data_stream_template_name);
+	snprintf(ctx->ds_uri, sizeof(ctx->ds_uri) - 1, "%s/_data_stream/%s", path, ctx->data_stream_name);
```
Use 4-white spaces indentations, please.

adiforluls replied:
Fixed

adiforluls commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -122,6 +122,17 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
+			}
+		}
+	}
+
```
My bad. Fixed now.

adiforluls commented on plugins/out_opensearch/opensearch.h:
```diff
@@ -89,6 +93,16 @@ struct flb_opensearch {
+	char *template_file;
+    int template_overwrite;
+	struct flb_record_accessor *ra_data_stream;
+
```
Fixed

adiforluls commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -336,38 +339,50 @@ static int opensearch_format(struct flb_config *config,
-                                         index);
+
+    cond1 = ctx->data_stream_mode == FLB_TRUE && ctx->ra_data_stream == NULL;
+    cond2 = ctx->logstash_format == FLB_FALSE && ctx->ra_index == NULL;
```
Fixed

adiforluls commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -336,38 +339,50 @@ static int opensearch_format(struct flb_config *config,
-                                         ctx->action,
-                                         index);
+
+    cond1 = ctx->data_stream_mode == FLB_TRUE && ctx->ra_data_stream == NULL;
```
Fixed

adiforluls commented:
@cosmo0920 forgot that I had edited other files and they also had weird indents. All should be fixed now.

cosmo0920 commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -29,6 +29,7 @@
 #include <fluent-bit/flb_ra_key.h>
 #include <fluent-bit/flb_log_event_decoder.h>
 #include <msgpack.h>
+#include <fluent-bit/flb_file.h>
```
`<fluent-bit/xxx.h> should be put in the similar headers. I mean, `<fluent-bit/flb_file.h>` should be better to put after `<fluent-bit/flb_log_event_decoder.h>`.

cosmo0920 commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -146,6 +157,8 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
         snprintf(ctx->uri, sizeof(ctx->uri) - 1, "%s/_bulk", path);
     }
 
+    snprintf(ctx->template_uri, sizeof(ctx->template_uri) - 1, "%s/_index_template/%s", path, ctx->data_stream_template_name);
+    snprintf(ctx->ds_uri, sizeof(ctx->ds_uri) - 1, "%s/_data_stream/%s", path, ctx->data_stream_name);
```
What is caused by `NULL` where are ctx->data_stream_template_name and ctx->data_stream_name?
This string operations should be proceeded only if they are specified or enabled data stream mode. What do you think?

adiforluls replied:
By default data_stream_name will be set to `fluent-bit` and data_stream_template_name will be `fluent-bit_template` when they're not provided. Do you think I should explicitly handle this? Something like:
```
if (ctx->data_stream_mode == FLB_TRUE && ctx->data_stream_name != NULL && ctx->data_stream_template_name != NULL)
...
```
My thinking was that the defaults for these values are non-null so we should be good in any case. Let me know if I should add handling here.

cosmo0920 replied:
Sorry, my bad. Your patch is correct.

cosmo0920 reviewed (commented):
Looks good for me.
@leonardo-albertovich Could you start to take a look for this PR? Logic for handling Data Streams seems good for me.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -336,38 +339,50 @@ static int opensearch_format(struct flb_config *config,
-                                         ctx->action,
-                                         index);
+
+    enabled_data_stream_no_ra = ctx->data_stream_mode == FLB_TRUE && ctx->ra_data_stream == NULL;
```
Please expand these two inlines conditionals

adiforluls replied:
Didn't get you, are you suggesting that I do something like:
```
if (enabled_data_stream_no_ra && ctx->generate_id == NULL
    || disabled_logstash_fmt_no_ra && ctx->generate_id == NULL) {
...
```
Let me know what you mean

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -304,6 +306,7 @@ static int opensearch_format(struct flb_config *config,
     flb_sds_t j_index;
     struct flb_log_event_decoder log_decoder;
     struct flb_log_event log_event;
+    int enabled_data_stream_no_ra, disabled_logstash_fmt_no_ra;
```
Please break these declarations in individual lines.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -381,6 +396,8 @@ static int opensearch_format(struct flb_config *config,
         flb_time_get(&tms);
     }
 
+    // TODO: Add timestamp to logs if data_stream_mode is true
```
Please use `/**/` in your comments.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -445,7 +462,41 @@ static int opensearch_format(struct flb_config *config,
+                    index = ra_index;
+                }
+
+                if (ctx->suppress_type_name) {
```
It seems like this block is duplicated and could be extracted into a small static function, could you?

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -853,12 +904,149 @@ static int opensearch_error_check(struct flb_opensearch *ctx,
+        return FLB_OS_RETRY;
+    }
+
+    template = flb_file_read(ctx->template_file);
```
`flb_file_read` can return `NULL`, please add the appropriate result verification conditional.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -853,12 +904,149 @@ static int opensearch_error_check(struct flb_opensearch *ctx,
+        return FLB_OS_RETRY;
+    }
+
+    c = flb_http_client(u_conn, FLB_HTTP_GET, ctx->template_uri,
```
`flb_http_client` can return `NULL`, please add the appropriate check here and in line 932.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -853,12 +904,149 @@ static int opensearch_error_check(struct flb_opensearch *ctx,
+
+}
+
+static int create_index_template(struct flb_opensearch *ctx)
```
Please break the lines that are longer than 80 characters in this function.

adiforluls commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -122,6 +122,17 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
 
+    if(ctx->data_stream_name) {
+        if(strchr(ctx->data_stream_name, '$')) {
+            ctx->ra_data_stream = flb_ra_create(ctx->data_stream_name, FLB_TRUE);
```
So if there's a problem in creating `ra_data_stream`, we destroy the context `ctx`. Will I have to explicitly destroy `ra_data_stream` that's part of `ctx`? 

It's the same as what's being done for ra_index as well (Line 116 in the same file). Let me know if this is incorrect and I'll fix this in both the places.

leonardo-albertovich replied:
What I'm saying is that whatever the case for the configuration context destruction is (ie. plugin exit), `flb_os_conf_destroy` does not release `ra_data_stream`.

You need to add that so it's properly disposed of.

leonardo-albertovich replied:
This requirement is still pending and critical to take this PR forward.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -336,38 +339,50 @@ static int opensearch_format(struct flb_config *config,
+    enabled_data_stream_no_ra = ctx->data_stream_mode == FLB_TRUE && ctx->ra_data_stream == NULL;
+    disabled_logstash_fmt_no_ra = ctx->logstash_format == FLB_FALSE && ctx->ra_index == NULL;
+
+    if ((enabled_data_stream_no_ra || disabled_logstash_fmt_no_ra) && ctx->generate_id == FLB_FALSE) {
```
```
if ((enabled_data_stream_no_ra == FLB_TRUE || 
     disabled_logstash_fmt_no_ra == FLB_TRUE ) && 
    ctx->generate_id == FLB_FALSE) {
```

adiforluls commented:
@leonardo-albertovich I've addressed all the comments and made the changes requested.

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -264,6 +264,46 @@ static int compose_index_header(struct flb_opensearch *ctx,
+    *ra_index = flb_ra_translate(rec,
+                                (char *) tag, tag_len,
+                                map, NULL);
+    if (!(*ra_index)) {
```
`*ra_index == NULL`

adiforluls replied:
Fixed

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -264,6 +264,46 @@ static int compose_index_header(struct flb_opensearch *ctx,
     return 0;
 }
 
+void create_ra_index(struct flb_opensearch *ctx,
```
Make this static

adiforluls replied:
Fixed

leonardo-albertovich commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -347,21 +388,36 @@ static int opensearch_format(struct flb_config *config,
 
-    if ((enabled_data_stream_no_ra || disabled_logstash_fmt_no_ra) && ctx->generate_id == FLB_FALSE) {
+    if ((enabled_data_stream_no_ra ==  FLB_TRUE ||
+    disabled_logstash_fmt_no_ra == FLB_TRUE) &&
```
indent this line and the next one appropriately so they are visually contextually consistent.

adiforluls replied:
Fixed

leonardo-albertovich commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -111,10 +111,11 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
         if (ctx->index) {
             /* do we have a record accessor pattern ? */
-            if (strchr(ctx->index, '$')) {
+            if (strchr(ctx->index, '$') != NULL) {
```
Isn't this supposed  to be `if (ctx->index[0] == '$')` to avoid false positives?

adiforluls replied:
No, I can form an index like foo-$record['key']

leonardo-albertovich commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -123,10 +124,11 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
 
     if(ctx->data_stream_name) {
-        if(strchr(ctx->data_stream_name, '$')) {
+        if(strchr(ctx->data_stream_name, '$') != NULL) {
```
Same as in 1114

adiforluls replied:
Answered above

adiforluls commented on plugins/out_opensearch/opensearch.c:
```diff
@@ -264,6 +264,46 @@ static int compose_index_header(struct flb_opensearch *ctx,
+                                (char *) tag, tag_len,
+                                map, NULL);
+    if (!(*ra_index)) {
+        flb_plg_warn(ctx->ins, "invalid data stream translation from record accessor pattern, default to static index");
```
Fixed

adiforluls commented on plugins/out_opensearch/os_conf.c:
```diff
@@ -111,17 +111,30 @@ struct flb_opensearch *flb_os_conf_create(struct flb_output_instance *ins,
+            ctx->ra_data_stream = flb_ra_create(ctx->data_stream_name, FLB_TRUE);
+            if(!ctx->ra_data_stream) {
+                flb_plg_error(ctx->ins, "invalid record accessor pattern set for 'data_stream_name' property");
+                flb_ra_destroy(ctx->ra_data_stream);
```
Oh, yeah, got it.

leonardo-albertovich commented:
Blessed be thy PR! This is ready to be merged @edsiper

leonardo-albertovich commented:
Thanks a lot for the contribution and cooperation @adiforluls!

cosmo0920 reviewed (approved):
:+1:

adiforluls commented:
Can we get this merged as soon as possible (possibly today)? It's an important feature for people who're using opensearch output so it'll be good to have this in a release. 
@edsiper

adiforluls commented:
Can a code owner look into this? I have two approvals from maintainers and need code-owner approval to merge. 
@edsiper or @PettitWesley

adiforluls commented:
This PR has been approved 3 weeks ago but is still pending merge, let me know if I'm missing some process
cc: @edsiper @PettitWesley @leonardo-albertovich @cosmo0920

PettitWesley commented:
@adiforluls I think the failing PR check unit test may be a blocker

adiforluls commented:
I see that only this check failed https://github.com/fluent/fluent-bit/actions/runs/5023119764/jobs/9008358596?pr=7371 and it doesn't look like the test ran? @PettitWesley or @leonardo-albertovich if you know what this test does

adiforluls commented:
@leonardo-albertovich or @cosmo0920 would you know if this failing test might be related? To me it looks like it did not even run so it is some unrelated regression. This PR is pending merge for about a month now so in case there's an issue I'd like to fix it ASAP, if not then would want to see this getting merged.

adiforluls commented:
Hi @edsiper, I received the approvals from the maintainers 4 months ago but this was never merged. Let me know what's holding up the merge. I've been running this change for months without any issues.

qxmips commented:
waiting to this functionality

ecerulm commented:
@adiforluls , the logs for the CI/CD jobs expired, could you retrigger them?

ecerulm commented:
Closes #8522
Closes #7079

ThoreKr commented:
This turned out to be a blocker for us for moving from fluentd to fluent-bit. Is there any progress here or alternatives to work with datastreams without a lot of manual interaction?

krushik commented:
@ThoreKr, you can set up an index template with the ‘Template type’ set to ‘Data streams’ and the required index pattern pre-configured on the OpenSearch side. Then you can use the regular write_operation 'create' in Fluent Bit with any index name that falls within that index pattern. This approach has been working well for us
